### PR TITLE
[GC] Backport async log

### PR DESCRIPTION
**Pull request by gaogao-mem**
Summary: Backport async log

Test Plan: test/hotspot/jtreg/gtest/AsyncLogGtest.java

Reviewed-by: D-D-H, zhengxiaolinX, leveretconey

Issue: #514

### CODE REVIEW

File: src/hotspot/share/logging/logFileStreamOutput.cpp
Comment (D-D-H) [inline src/hotspot/share/logging/logFileStreamOutput.cpp:1]: missing change

File: src/hotspot/share/logging/logDecorations.hpp [hunk 1]
```diff
@@ -27,27 +27,34 @@
...
+  const intx _tid;                // for "tid"
+  LogLevelType _level;            // for "level" (needs to be nonconst)
+  const LogTagSet& _tagset;       // for "tags"
+  // In debug mode we keep the decorators around for sanity checking when printing
+  DEBUG_ONLY(const LogDecorators& _decorators;)
 
-  jlong java_millis();
-  void create_decorations(const LogDecorators& decorators);
+  static const char* _host_name;
+  static const int _pid;
```
Comment (D-D-H) [inline src/hotspot/share/logging/logDecorations.hpp:48]: inconsistent change

File: .gitignore [hunk 2]
```diff
@@ -13,3 +13,11 @@ test/nashorn/lib
...
 **/JTreport/**
 **/JTwork/**
+**/*.iml
+**/.idea/**
+**/CMakeLists.txt
+**/cmake-build-debug/**
+.repo/
+graalvm/
+mx/
+/src/utils/LogCompilation/target/
```
Comment (D-D-H) [inline .gitignore:23]: 8229517: unnecessary change

File: src/hotspot/share/logging/logConfiguration.hpp [hunk 3]
```diff
@@ -123,6 +124,11 @@ class LogConfiguration : public AllStatic {
 
   // Rotates all LogOutput
   static void rotate_all_outputs();
+
+  static bool is_async_mode() { return _async_mode; }
+  static void set_async_mode(bool value) {
+    _async_mode = value;
+  }
```
Comment (D-D-H) [inline src/hotspot/share/logging/logConfiguration.hpp:131]: 8229517: missing changes in [logDecorators.hpp](https://github.com/openjdk/jdk/commit/41185d38f21e448370433f7e4f1633777cab6170#diff-d2a0b96dcf40197d99b4d0c795d2f9802aa5a4de4a9a7a1b85666d687b8a783e)

File: src/hotspot/share/logging/logFileOutput.hpp [hunk 4]
```diff
@@ -83,6 +83,7 @@ class LogFileOutput : public LogFileStreamOutput {
   LogFileOutput(const char *name);
   virtual ~LogFileOutput();
   virtual bool initialize(const char* options, outputStream* errstream);
+  int write_blocking(const LogDecorations& decorations, const char* msg);
```
Comment (D-D-H) [inline src/hotspot/share/logging/logFileOutput.hpp]: 8229517: wrong position

File: src/hotspot/share/runtime/thread.cpp [hunk 5]
```diff
@@ -368,6 +369,9 @@ void Thread::record_stack_base_and_size() {
 void Thread::register_thread_stack_with_NMT() {
   MemTracker::record_thread_stack(stack_end(), stack_size());
 }
+void Thread::unregister_thread_stack_with_NMT() {
```
Comment (D-D-H) [inline src/hotspot/share/runtime/thread.cpp:372]: 8229517: why is the method needed

File: src/hotspot/share/logging/logAsyncWriter.cpp [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    _stats.iterate(&dropped_counters_iter);
+    own_io = _io_sem.trywait();
+  }
+
+  LinkedListIterator<AsyncLogMessage> it(logs.head());
+  if (!own_io) {
+    _io_sem.wait();
+  }
+
+  while (!it.has_next()) {
```
Comment (D-D-H) [inline src/hotspot/share/logging/logAsyncWriter.cpp:137]: It is is_empty here in upstream. We'd betterr keep it same if the cost is trivial.

**Review by D-D-H — COMMENTED** [review]
8267926: missing change in test/hotspot/jtreg/ProblemList.txt

File: src/hotspot/share/logging/logFileOutput.cpp
Comment (D-D-H) [inline src/hotspot/share/logging/logFileOutput.cpp:1]: 8268165: missing change in LogFileOutput::rotate

File: src/hotspot/share/logging/logAsyncWriter.cpp
Comment (D-D-H) [inline src/hotspot/share/logging/logAsyncWriter.cpp:1]: 8267952: missing change in AsyncLogWriter::enqueue_locked

File: src/hotspot/share/logging/logAsyncWriter.cpp [hunk 7]
```diff
@@ -128,24 +126,28 @@ void AsyncLogWriter::write() {
...
       e->output()->write_blocking(e->decorations(), msg);
       os::free(msg);
+    } else if (e->output() == NULL) {
+      // This is a flush token. Record that we found it and then
+      // signal the flushing thread after the loop.
+      req++;
     }
   }
-  _io_sem.signal();
+  if (req > 0) {
```
Comment (D-D-H) [inline src/hotspot/share/logging/logAsyncWriter.cpp:150]: 8267952: misling a newline

File: src/hotspot/share/logging/logAsyncWriter.hpp [hunk 8]
```diff
@@ -121,17 +121,16 @@ typedef KVHashtable<LogFileOutput*, uint32_t, mtLogging> AsyncLogMap;
 // initialize() is called once when JVM is initialized. It creates and initializes the singleton instance of AsyncLogWriter.
 // Once async logging is established, there's no way to turn it off.
 //
-// instance() is MT-safe and returns the pointer of the singleton instance if and only if async logging is enabled and has well
+// instance() is MT-safe and returns the pointer of the singleton instance if and only if async logging is enabled and has successfully
```
Comment (D-D-H) [inline src/hotspot/share/logging/logAsyncWriter.hpp:133]: 8267952: inconsistent change

File: src/hotspot/share/runtime/perfData.cpp [hunk 9]
```diff
@@ -279,6 +279,10 @@ void PerfDataManager::destroy() {
   _has_PerfData = false;
   os::naked_short_sleep(1);  // 1ms sleep to let other thread(s) run
 
+  log_debug(perf, datacreation)("Total = %d, Sampled = %d, Constants = %d",
+                                _all->length(), _sampled == NULL ? 0 : _sampled->length(),
+                                _constants == NULL ? 0 : _constants->length());
+
```
Comment (D-D-H) [inline src/hotspot/share/runtime/perfData.cpp:285]: 8271353: This patch seems unnecessary.

File: src/hotspot/share/logging/logAsyncWriter.cpp [hunk 10]
```diff
@@ -43,7 +44,7 @@ class AsyncLogWriter::AsyncLogLocker : public StackObj {
 void AsyncLogWriter::enqueue_locked(const AsyncLogMessage& msg) {
   if (_buffer.size() >= _buffer_max_size)  {
     bool p_created;
-    uint32_t* counter = _stats.add_if_absent(msg.output(), 0, &p_created);
+    uint32_t* counter = _stats.put_if_absent(msg.output(), 0, &p_created);
```
Comment (D-D-H) [inline src/hotspot/share/logging/logAsyncWriter.cpp:47]: 8267517: this change doesn't exist in 8267517

File: src/hotspot/share/logging/logAsyncWriter.cpp [hunk 11]
```diff
@@ -78,7 +79,7 @@ void AsyncLogWriter::enqueue(LogFileOutput& output, LogMessageBuffer::Iterator m
 AsyncLogWriter::AsyncLogWriter()
   : _flush_sem(0), _lock(), _data_available(false),
     _initialized(false),
-    _stats(17 /*table_size*/),
+    _stats(),
```
Comment (D-D-H) [inline src/hotspot/share/logging/logAsyncWriter.cpp:82]: 8267517: this change doesn't exist in 8267517

File: src/hotspot/share/logging/logFileStreamOutput.cpp [hunk 12]
```diff
@@ -77,12 +78,10 @@ int LogFileStreamOutput::write_decorations(const LogDecorations& decorations) {
 class FileLocker : public StackObj {
 private:
   FILE *_file;
-
```
Comment (D-D-H) [inline src/hotspot/share/logging/logFileStreamOutput.cpp]: 8267517: this change doesn't exist in 8267517

File: src/hotspot/share/logging/logFileStreamOutput.cpp [hunk 13]
```diff
@@ -77,12 +78,10 @@ int LogFileStreamOutput::write_decorations(const LogDecorations& decorations) {
...
 public:
   FileLocker(FILE *file) : _file(file) {
     os::flockfile(_file);
   }
-
```
Comment (D-D-H) [inline src/hotspot/share/logging/logFileStreamOutput.cpp]: 8267517: this change doesn't exist in 8267517

**Review by D-D-H — COMMENTED** [review]
The backport overall looks good.

I left some comments with related bug ID as prefix.

File: src/hotspot/share/logging/logAsyncWriter.cpp (hunk 6 above)
Comment (gaogao-mem) [reply src/hotspot/share/logging/logAsyncWriter.cpp:137]: retain and add annotation

File: src/hotspot/share/runtime/perfData.cpp (hunk 9 above)
Comment (gaogao-mem) [reply src/hotspot/share/runtime/perfData.cpp:285]: deleted

File: src/hotspot/os/windows/os_windows.cpp [hunk 14]
```diff
@@ -5456,6 +5456,55 @@ void Parker::unpark() {
   SetEvent(_ParkEvent);
 }
 
+// Platform Monitor implementation
+
+os::PlatformMonitor::PlatformMonitor() {
```
Comment (zhengxiaolinX) [inline src/hotspot/os/windows/os_windows.cpp:5461]: These are interesting adaptions from 8210832 and 8228857, but okay.

**Review by zhengxiaolinX — APPROVED** [review]
Generally lgtm.

**Review by leveretconey — COMMENTED** [review]
looks good to me

**Review by leveretconey — APPROVED** [review]

**Review by D-D-H — APPROVED** [review]
